keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+    a _different_ configurable bit on the packet mark. `KUBE-FIREWALL`
+    checks the packet mark and calls `-j DROP` on the packets that
+    were previously marked for dropping.
+
+    These chains have always been created by kubelet, but were only
+    ever used by kube-proxy.
+
+The reason that these helper chains exist is because most of
+kube-proxy's service-processing logic occurs in subchains of the
+`OUTPUT` and `PREROUTING` chains of the `nat` table (because those are
+the only places you can call the `DNAT` target from, to redirect
+packets from a service IP to a pod IP), but neither masquerading nor
+dropping packets can occur at that point in IPTables processing:
+dropping packets can only occur from chains in the `filter` table,
+while masquerading can only occur from the `POSTROUTING` chain of the
+`nat` table (because the kernel can't know what IP to masquerade the
+packet to until after it has decided what interface it will send it
+out on, which necessarily can't happen until after it knows if it is
+going to DNAT it). To work around this, when Kubernetes wants to
+masquerade or drop a packet, it just marks it as needing to be
+masqueraded or dropped later, and then one of the later chains handles
+the actual masquerading/dropping.
+
+This approach was not _necessary_; kube-proxy could have just
+duplicated its matching logic between multiple IPTables chains, eg,
+having a rule saying "a packet sent to `192.168.0.11:80` should be
+redirected to `10.0.20.18:8080`" in one chain and a rule saying "a
+packet whose original pre-NAT destination was `192.168.0.11:80` should
+be masqueraded" in another.
+
+But using `KUBE-MARK-MASQ` allows kube-proxy to keep its logic more
+efficient, compact, and well-organized than it would be otherwise; it
+can just create a pair of adjacent rules saying "a packet sent to
+`192.168.0.11:80` should be redirected to `10.0.20.18:8080`, and
+should also be masqueraded".
+
+#### Eliminating `KUBE-MARK-DROP`
+
+In theory, `KUBE-MARK-DROP` is useful for the same reason as
+`KUBE-MARK-MASQ`, although in practice, it turns out to be much less
+necessary. Kube-proxy currently drops packets in two cases:
+
+  - When using the `LoadBalancerSourceRanges` feature on a service,
+    any packet to that service that comes from outside the valid
+    source IP ranges is dropped.
+
+  - When using `Local` external traffic policy, if a connection to a
+    service arrives on a node that has no local endpoints for that
+    service, it is dropped.
+
+In the `LoadBalancerSourceRanges` case, it would not be difficult to
+avoid using `KUBE-MARK-DROP`. The current rule logic creates a
+per-service `KUBE-FW-` chain that checks against each allowed source
+range and forwards traffic from those ranges to the `KUBE-SVC-` or
+`KUBE-XLB-` chain for the service. At the end of the chain, it calls
+`KUBE-MARK-DROP` on any unmatched packets.
+
+To do this without `KUBE-MARK-DROP`, we simply remove the
```

Comment:
I find the fact that the processing logic is in one place is convenient for comprehension, but I guess we can free up one bit in the mark...